### PR TITLE
Implement changes to MSC2285 (private read receipts)

### PR DESCRIPTION
**Pull request by SimonBrandner**
Type: enhancement
Implement [changes](https://github.com/matrix-org/matrix-spec-proposals/pull/2285/commits/4a77139249c2e830aec3c7d6bd5501a514d1cc27) to [MSC2285 (hidden read receipts)](https://github.com/matrix-org/matrix-spec-proposals/pull/2285)
Requires https://github.com/matrix-org/matrix-react-sdk/pull/7993
Requires https://github.com/matrix-org/synapse/pull/12168

<hr>

The changes in this PR shouldn't affect any clients which don't currently use the opts params.

## `sendReadReceipt(event: MatrixEvent, receiptType = ReceiptType.Read, callback?: Callback)`

The `opts` param has been replaced with `receiptType: ReceiptType`. By default this is `ReceiptType.Read` and any other value is  _experimental_.

## `setRoomReadMarkers(roomId: string, rmEventId: string, rrEvent?: MatrixEvent, rpEvent?: MatrixEvent)`

The `opts` param has been replaced with `rpEvent?: MatrixEvent`. This param is optional and _experimental_. It allows you to specify a private read receipt not visible to other users

## `setRoomReadMarkersHttpRequest(roomId: string, rmEventId: string, rrEventId: string, rpEventId: string)`

The `opts` param has been replaced with `rpEventId: string`. This param is optional and _experimental_. It allows you to specify a private read receipt not visible to other users

---
Here's what your changelog entry will look like:

## 🚨 BREAKING CHANGES
 * Implement changes to MSC2285 (private read receipts) ([\#2221](https://github.com/matrix-org/matrix-js-sdk/pull/2221)).

### CODE REVIEW

File: src/sync-accumulator.ts [hunk 1]
```diff
@@ -433,13 +435,24 @@ export class SyncAccumulator {
                 // of a hassle to work with. We'll inflate this back out when
                 // getJSON() is called.
                 Object.keys(e.content).forEach((eventId) => {
-                    if (!e.content[eventId]["m.read"]) {
+                    if (!e.content[eventId][ReceiptType.Read] && !e.content[eventId][ReceiptType.ReadPrivate]) {
                         return;
                     }
-                    Object.keys(e.content[eventId]["m.read"]).forEach((userId) => {
+                    const read = e.content[eventId][ReceiptType.Read];
+                    read && Object.keys(read).forEach((userId) => {
```
Comment (t3chguy) [inline src/sync-accumulator.ts:442]: not a fan of this style

File: src/sync-accumulator.ts [hunk 2]
```diff
@@ -433,13 +435,24 @@ export class SyncAccumulator {
...
                         // clobber on user ID
                         currentData._readReceipts[userId] = {
-                            data: e.content[eventId]["m.read"][userId],
+                            data: e.content[eventId][ReceiptType.Read][userId],
+                            type: ReceiptType.Read,
+                            eventId: eventId,
+                        };
+                    });
+                    const readPrivate = e.content[eventId][ReceiptType.ReadPrivate];
+                    readPrivate && Object.keys(readPrivate).forEach((userId) => {
```
Comment (t3chguy) [inline src/sync-accumulator.ts:451]: same as above

File: src/client.ts [hunk 3]
```diff
@@ -4561,11 +4555,23 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
                 throw new Error(`Cannot set read receipt to a pending event (${rrEventId})`);
             }
             if (room) {
-                room.addLocalEchoReceipt(this.credentials.userId, rrEvent, "m.read");
+                room.addLocalEchoReceipt(this.credentials.userId, rrEvent, ReceiptType.Read);
+            }
+        }
+
+        // Add the optional private RR update, do local echo like `sendReceipt`
+        let rpEventId;
```
Comment (t3chguy) [inline src/client.ts:4563]:
```suggestion
        let rpEventId: string;
```

File: src/client.ts [hunk 4]
```diff
@@ -4561,11 +4555,23 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
...
+        if (rpEvent) {
+            rpEventId = rpEvent.getId();
+            if (room && room.hasPendingEvent(rpEventId)) {
```
Comment (t3chguy) [inline src/client.ts:4566]:
```suggestion
            if (room?.hasPendingEvent(rpEventId)) {
```

File: src/client.ts [hunk 5]
```diff
@@ -4561,11 +4555,23 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
...
+                throw new Error(`Cannot set read receipt to a pending event (${rpEventId})`);
+            }
+            if (room) {
+                room.addLocalEchoReceipt(this.credentials.userId, rpEvent, ReceiptType.ReadPrivate);
             }
```
Comment (t3chguy) [inline src/client.ts:4571]:
```suggestion
            room?.addLocalEchoReceipt(this.credentials.userId, rpEvent, ReceiptType.ReadPrivate);
```

**Review by t3chguy — APPROVED** [review]
Looks sane otherwise

File: src/models/room.ts [hunk 6]
```diff
@@ -2320,8 +2321,10 @@ export class Room extends TypedEventEmitter<EmittedEvents, RoomEventHandlerMap>
         });
     }
 
-    public getReadReceiptForUserId(userId: string, ignoreSynthesized = false): IWrappedReceipt | null {
-        const [realReceipt, syntheticReceipt] = this.receipts["m.read"]?.[userId] ?? [];
+    public getReadReceiptForUserId(
```
Comment (t3chguy) [inline src/models/room.ts:2324]: Any chance this can get a jsdoc comment?

File: src/models/room.ts [hunk 7]
```diff
@@ -2339,8 +2342,22 @@ export class Room extends TypedEventEmitter<EmittedEvents, RoomEventHandlerMap>
...
+        if (publicReadReceipt?.eventId && privateReadReceipt?.eventId) {
+            comparison = timelineSet.compareEventOrdering(publicReadReceipt?.eventId, privateReadReceipt?.eventId);
+        }
+
+        // The public receipt is more likely to drift out of date so the private
+        // one has precedence
+        if (!comparison) return privateReadReceipt?.eventId ?? publicReadReceipt?.eventId ?? null;
+
+        // If public read receipt is older, return the private one
+        return (comparison < 0) ? privateReadReceipt?.eventId : publicReadReceipt?.eventId;
```
Comment (t3chguy) [inline src/models/room.ts:2360]: Can we not use the `ts` field of the receipt?

Comment (SimonBrandner) [reply src/models/room.ts:2367]: Technically we could have an older event `A` and newer event `B`, send an `m.read.private` receipt to event `B` and then send `m.read` receipt to event `A`. In this case, we would still want to return event `B` but the receipt pointing to event `A` would be the one with the newer `ts`, so event `A` would be returned which would be wrong

Comment (t3chguy) [reply src/models/room.ts:2367]: I mean for the case of when we can't do the comparison due to not having both events in a single timeline

File: src/models/room.ts [hunk 8]
```diff
@@ -2339,8 +2349,22 @@ export class Room extends TypedEventEmitter<EmittedEvents, RoomEventHandlerMap>
...
      */
     public getEventReadUpTo(userId: string, ignoreSynthesized = false): string | null {
-        const readReceipt = this.getReadReceiptForUserId(userId, ignoreSynthesized);
-        return readReceipt?.eventId ?? null;
+        const timelineSet = this.getUnfilteredTimelineSet();
+        const publicReadReceipt = this.getReadReceiptForUserId(userId, ignoreSynthesized, ReceiptType.Read);
+        const privateReadReceipt = this.getReadReceiptForUserId(userId, ignoreSynthesized, ReceiptType.ReadPrivate);
+
+        // If we have both, compare them
+        let comparison;
```
Comment (t3chguy) [inline src/models/room.ts:2357]: type please :)

File: src/models/room.ts (hunk 7 above)
Comment (SimonBrandner) [reply src/models/room.ts]: Ah, I see, yep, we can do that!

**Review by t3chguy — APPROVED** [review]

**Pull request merged**